src/analyses/variable-sensitivity/three_way_merge_abstract_interpreter.cpp

```diff
@@ -0,0 +1,255 @@
+      calling_function_id,
+      p_call,
+      callee_function_id,
+      l_callee_start,
+      ai_history_baset::
+        no_caller_history, // Not needed as p_call already has the info
+      ns,
+      catch_working_set);
+
+    // do we need to do/re-do the fixedpoint of the body?
+    if(new_data)
+      fixedpoint(
+        get_next(catch_working_set), // Should be p_callee_start...
+        callee_function_id,
+        callee,
+        goto_functions,
+        ns);
+  }
+
+  // Now we can give histories for the return part
+
+  // Find the histories at the end of the function
+  auto traces = storage->abstract_traces_before(l_callee_end);
+
+  bool new_data = false; // Whether we have changed a domain in the caller
+
+  // As with recursive-interprocedural, there are potentially multiple histories
+  // at the end, or maybe none.  Only some of these will be 'descendents' of
+  // p_call_site and p_callee_start
+  for(auto p_callee_end : *traces)
+  {
+    // First off, is it even reachable?
+    const statet &s_callee_end = get_state(p_callee_end);
+
+    if(s_callee_end.is_bottom())
+      continue; // Unreachable in callee -- no need to merge
+
+    // Can it return to p_call_site?
+    auto return_step = p_callee_end->step(
+      l_return_site,
+      *(storage->abstract_traces_before(l_return_site)),
+      p_call_site); // Because it is a return edge!
+    if(return_step.first == ai_history_baset::step_statust::BLOCKED)
+      continue; // Can't return -- no need to merge
+
+    // The fourth history!
+    trace_ptrt p_return_site = return_step.second;
+
+    const std::unique_ptr<statet> ptr_s_callee_end_copy(
+      make_temporary_state(s_callee_end));
+    auto tmp =
+      dynamic_cast<variable_sensitivity_domaint *>(&(*ptr_s_callee_end_copy));
+    INVARIANT(tmp != nullptr, "Three-way merge requires domain support");
+    variable_sensitivity_domaint &s_working = *tmp;
+
+    // Apply transformer
+    // This is for an end_function instruction which normally doesn't do much
+    // but in VSD it does, so this cannot be omitted.
+    s_working.transform(
+      callee_function_id,
+      l_callee_end,
+      calling_function_id,
+      l_return_site,
+      *this,
+      ns);
+
+    // TODO : this is probably needed to avoid three_way_merge modifying one of
+    // it's arguments as it goes.  A better solution would be to refactor
+    // merge_three_way_function_return.
+    const std::unique_ptr<statet> ptr_s_working_copy(
+      make_temporary_state(s_working));
+
+    s_working.merge_three_way_function_return(
+      get_state(p_call_site),
+      get_state(p_callee_start),
+      *ptr_s_working_copy,
+      ns);
+
+    if(
+      merge(s_working, p_callee_end, p_return_site) ||
+      (return_step.first == ai_history_baset::step_statust::NEW &&
+       !s_working.is_bottom()))
+    {
+      put_in_working_set(working_set, p_return_site);
+      new_data = true;
+    }
+  }
+
+  return new_data;
+}
+
+#if 0
```

Comment:
Please document why this is commented out, or remove.

Reply:
Ooops :-\  Removed!